tools/convert/main.cpp

```diff
@@ -230,74 +255,88 @@ int main(int argc, const char** argv)
-            json_consumer.StartFile(out_file_handle);
+            json_consumer.Initialize(&json_writer, vulkan_version);
+            json_writer.StartStream(&out_stream);
 
             // If CONVERT_EXPERIMENTAL_D3D12 was set, then add DX12 consumer/decoder
 #ifdef CONVERT_EXPERIMENTAL_D3D12
-            gfxrecon::decode::Dx12AsciiConsumer dx12_ascii_consumer;
-            gfxrecon::decode::Dx12Decoder       dx12_decoder;
+            Dx12JsonConsumer              dx12_json_consumer;
+            gfxrecon::decode::Dx12Decoder dx12_decoder;
 
-            dx12_decoder.AddConsumer(&dx12_ascii_consumer);
+            dx12_decoder.AddConsumer(&dx12_json_consumer);
             file_processor.AddDecoder(&dx12_decoder);
             auto dx12_json_flags = output_format == JsonFormat::JSON ? gfxrecon::util::kToString_Formatted
                                                                      : gfxrecon::util::kToString_Unformatted;
-            dx12_ascii_consumer.Initialize(out_file_handle, dx12_json_flags);
+            dx12_json_consumer.Initialize(&json_writer);
 #endif
 
             while (success)
             {
                 success = file_processor.ProcessNextFrame();
                 if (success && file_per_frame)
                 {
-                    json_consumer.EndFile();
-                    gfxrecon::util::platform::FileClose(out_file_handle);
+                    json_writer.EndStream();
+                    const auto close_result = gfxrecon::util::platform::FileClose(out_file_handle);
+                    if (close_result != 0)
+                    {
+                        GFXRECON_LOG_ERROR("Failed to close file: '%s'.", json_filename.c_str());
+                        ret_code = 1;
+                    }
                     json_filename = gfxrecon::util::filepath::InsertFilenamePostfix(
                         output_filename, +"_" + FormatFrameNumber(file_processor.GetCurrentFrameNumber()));
-                    gfxrecon::util::platform::FileOpen(&out_file_handle, json_filename.c_str(), "w");
-                    success = out_file_handle != nullptr;
+                    const auto open_result =
+                        gfxrecon::util::platform::FileOpen(&out_file_handle, json_filename.c_str(), "w");
+                    success = open_result == 0 && out_file_handle != nullptr;
                     if (success)
                     {
-                        json_consumer.StartFile(out_file_handle);
+                        out_stream.Reset(out_file_handle);
+                        json_writer.StartStream(&out_stream);
                     }
                     else
                     {
-                        GFXRECON_LOG_ERROR("Failed to create file: '%s'.", json_filename.c_str());
+                        GFXRECON_LOG_ERROR("Failed to open file: '%s'.", json_filename.c_str());
                         ret_code = 1;
                     }
```

Comment:
Everything will probably crash and burn while outputting the next frame to a closed file so in addition, maybe one of 
* Exit here.
* Plumb in some sort of NullOutStream instead of the file stream which drops the frame.
* Reset the FileOutStream to a null file handle and make sure output ops are robust to a null handle.